examples/authentication/client/src/main.rs

```diff
@@ -0,0 +1,194 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! Example client authenticating to Oak using OpenID Connect.
+//!
+//! This example uses the Google Identity Platform.
+//! https://developers.google.com/identity/
+
+use futures::channel::oneshot::{self, Sender};
+use futures_util::future;
+use hyper::{service::Service, Body, Request, Response, Server, StatusCode};
+use log::{info, warn};
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+use structopt::StructOpt;
+use url::{form_urlencoded, Url};
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "OpenID Connect Client Example.")]
+pub struct Opt {
+    #[structopt(
+        long,
+        help = "Address of the Oak application to connect to.",
+        default_value = "127.0.0.1:8080"
+    )]
+    address: String,
+    #[structopt(
+        long,
+        help = "Address to listen on for the OAuth redirect.",
+        default_value = "127.0.0.1:8089"
+    )]
+    redirect_address: String,
+    #[structopt(long, help = "Path to the PEM-encoded CA root certificate.")]
+    ca_cert: Option<String>,
+    // TODO(#886): Retrieve client_id from server, rather than flag.
+    #[structopt(long, help = "The OAuth client ID.")]
+    client_id: String,
+}
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    env_logger::init();
+    let opt = Opt::from_args();
+    let redirect_address = opt.redirect_address.parse()?;
+    info!("Listening for redirect on {}", &redirect_address);
+
+    info!("Opening Auth request in Browser");
+    let request_uri = get_request(opt);
+    info!("URI: {}", &request_uri);
+    let (sender, receiver) = oneshot::channel();
+    let result_sender = Arc::new(Mutex::new(ResultSender {
+        sender: Some(sender),
+        result: None,
+    }));
+    let maker = MakeHandler {
+        result_sender: result_sender.clone(),
+    };
+
+    let mut runner = tokio::runtime::Runtime::new().unwrap();
+    let task = runner.spawn(async move {
+        Server::bind(&redirect_address)
+            .serve(maker)
+            .with_graceful_shutdown(async move {
+                receiver.await.unwrap();
+            })
+            .await
+    });
+    open::that(request_uri)?;
+    runner.block_on(task)??;
+
+    let result = result_sender.lock().unwrap();
+    let code = result.result.as_ref().unwrap().as_ref().ok().unwrap();
+    info!("Received code: {}", code);
+    Ok(())
+}
+
+fn get_request(opt: Opt) -> String {
```

Comment:
Rename to something more specific and / or add some documentation.

Reply:
Done.